### DISCUSSION TITLE
Implementing ADKernel for chemical potential equations

### DISCUSSION THREAD

**Discussion by ahzeeshan** [idaholab/moose#28591] [Q&A Modules: Phase field]
I am trying to implement an ADKernel for satisfying equilibrium conditions (pointwise) for chemical potentials of two species: $\mu_a$ and $\mu_b$. The conditions are $\mu_a + \mu_b = 0$ and $\mu_a = \mu_{target}$ (constant). The concentrations of species are $c_a$ and $c_b$. The code has three variables: $\phi, c_a, c_b$. 

Note that $\mu_a$ and $\mu_b$ are themselves defined in terms of $c_a, c_b$ and $\phi$. Let $\mu_a = f(c_a, c_b, \phi)$, $\mu_b = g(c_a, c_b, \phi)$. 

**Implementation**
*Material properties*
I implemented the chemical potential functions $\mu_a = f(c_a, c_b, \phi)$ and $\mu_b = g(c_a, c_b, \phi)$ as material properties using the `ADDerivativeParsedMaterial`, see below

```
  [./chempot_pos]
     type = ADDerivativeParsedMaterial
     property_name = 'mu_pos'
     coupled_variables = 'phi cpos cneg'
     material_property_names = 'mu0_pos kT zpos e'
     expression = 'mu0_pos + kT * log(cpos / (1-cpos)) + zpos * e * phi'
     derivative_order = 2
     outputs = exodus
  [../]
```
*Kernels*
The kernels are: a PDE for $\phi$; $\mu_a + \mu_b = 0$ for $c_b$ and $\mu_a = \mu_0$ for $c_a$. I implemented ADKernels for the two conditions by taking as input the material properties $\mu_a$, $\mu_b$, $\mu_0$ 
```
params.addRequiredParam<MaterialPropertyName>("mua", "The name of the first chemical potential");
...
_mua(getADMaterialProperty<Real>("mua")),
```
and setting the residuals as:
``return  (_mu[_qp] - _mutarget[_qp]) * _test[_i][_qp];``

``return  (_mua[_qp] + _mub[_qp]) * _test[_i][_qp];``

The kernels in the input file look like this
```
  [./chem_pot_eqbm]
    type = ChemPotEqual
    variable = cneg
    mua = mu_pos
    mub = mu_neg
    args = 'phi cneg'
  [../]
```

*Problem*
I am unable to achieve convergence. By running with `-snes_test_jacobian`, I found that my Jacobians are wrong. I am not entirely sure but maybe the problem stems from passing the $\mu$'s directly to the ADKernel as material properties instead of hard coding the functions $f$ and $g$ inside the kernel that make the dependence on $c_a, c_b$ and $\phi$ explicit?

I'd appreciate any help.

**Comment by GiudGiud** [comment]
Hello

The Jacobians should be correct when using AD without mistakes dropping the derivatives, so we should look for those.
This way of retrieving  material properties should be correct. Did you use this consistently?
```
_mua(getADMaterialProperty<Real>("mua")),
```

This expression will diverge near 0 and 1 `log(cpos / (1-cpos))`. Do you have a Bounds object maybe to make sure cpos does not approach the bounds?

>  maybe the problem stems from passing the μ 's directly to the ADKernel as material properties instead of hard coding the functions f and g inside the kernel that make the dependence on ca,cb and ϕ explicit?

no using AD properties should be fine for propagating the derivatives correctly from the residual to the Jacobian

**Comment by ahzeeshan** [comment]
Hi @GiudGiud 

Yes, I used this consistently in the kernels. I also added a bounds object but I got the same problem. Could you please look into my app here: [https://github.com/ahzeeshan/edl_solid/blob/main/test/tests/simple.i](https://github.com/ahzeeshan/edl_solid/blob/main/test/tests/simple.i)

**Reply by GiudGiud** [reply]
could you try something like `log( max(1e-12, cpos / (1-cpos) ) )` to prevent getting arbitrarily close to 0 in the log? The bounds don't prevent that

**Comment by ahzeeshan** [comment]
I tried this but the Jacobian is still off.
`||J - Jfd||_F/||J||_F = 0.000486969, ||J - Jfd||_F = 55.5544`

**Reply by GiudGiud** [reply]
bounding it will not help the Jacobian but it will prevent floating point errors.
Does it give you which coefficient is off? Using the Outputs/dofmap=true, we should look at which variable the Jacobian is not correct for
